p2p/p2p.go

```diff
@@ -71,115 +100,165 @@ func (p2p *P2P) Run() {
 		}
-		responder = request.Responder
-	case jsonrpc.QueryNumPeersRequest:
-		// Return the number of entries in the store.
-		entries := p2p.store.Entries()
-		response = jsonrpc.QueryNumPeersResponse{
-			NumPeers: entries,
-		}
-		responder = request.Responder
-	default:
-		p2p.logger.Panicf("unknown query request type: %T", request)
+		return nil
 	}
-
-	if responder == nil {
-		// This is a defensive check.
-		p2p.logger.Error("query responder channel is nil")
-	} else {
-		responder <- response
+	if err := p2p.store.InsertMultiAddress(multi.Addr(), multi); err != nil {
+		p2p.logger.Errorf("cannot add multi-address to store: %v", err)
+		return nil
+	}
+	if response.Error != nil {
+		p2p.logger.Warnf("received error in response: code = %v, message = %v, data = %v", response.Error.Code, response.Error.Message, string(response.Error.Data))
+		return nil
 	}
 
-	return nil
+	return response.Result
 }
 
-// randomPeers returns at most 5 random multi-addresses from the store.
-func (p2p *P2P) randomPeers(iter store.KVStoreIterator) []string {
-	// Retrieve all the Darknode multi-addresses in the store.
-	addresses := make([]string, 0, p2p.store.Entries())
-	for iter.Next() {
-		var value peer.MultiAddr
-		_, err := iter.KV(&value)
-		if err != nil {
-			p2p.logger.Errorf("cannot read multi-address using iterator: %v", err)
-			continue
-		}
-		addresses = append(addresses, value.Value())
+// handleQueryPeers retrieves at most 5 random multi-addresses from the store.
+func (p2p *P2P) handleQueryPeers(request jsonrpc.QueryPeersRequest) jsonrpc.Response {
+	addresses := p2p.randomPeers()
+	response := jsonrpc.QueryPeersResponse{
+		Peers: addresses,
 	}
+	return response
+}
+
+func (p2p *P2P) randomPeers() []string {
+	// Retrieve all the Darknode multi-addresses in the store.
+	addresses := p2p.store.MultiAddresses()
 
 	// Shuffle the list.
 	rand.Shuffle(len(addresses), func(i, j int) {
 		addresses[i], addresses[j] = addresses[j], addresses[i]
 	})
 
 	// Return at most 5 addresses.
-	length := p2p.multiAddrCount
-	if len(addresses) < p2p.multiAddrCount {
+	length := p2p.peerCount
+	if len(addresses) < p2p.peerCount {
 		length = len(addresses)
 	}
 
 	return addresses[:length]
 }
+
+// handleQueryNumPeers retrieves the number of multi-addresses in the store.
+func (p2p *P2P) handleQueryNumPeers(request jsonrpc.QueryNumPeersRequest) jsonrpc.Response {
+	// Return the number of entries in the store.
+	entries := p2p.store.MultiAddressEntries()
+	response := jsonrpc.QueryNumPeersResponse{
+		NumPeers: entries,
+	}
+	return response
+}
+
+// handleQueryStats retrieves the stats for the given Darknode address from the store.
+func (p2p *P2P) handleQueryStats(request jsonrpc.QueryStatsRequest) jsonrpc.Response {
+	if request.DarknodeID == "" {
+		// If no Darknode ID is provided, return the stats for the Lightnode.
+		cpus, err := p2p.health.CPUs()
+		if err != nil {
+			return nil
```

Comment:
In this case, we're returning a nil response and we didn't do any check before writing to the responder channel, this will cause the server panicing.